src/services/refactors/convertToOptionalChainExpression.ts

```diff
@@ -111,35 +107,63 @@ namespace ts.refactor.convertToOptionalChainExpression {
         const occurrences: (PropertyAccessExpression | Identifier)[] = [];
         while (isBinaryExpression(expression) && expression.operatorToken.kind === SyntaxKind.AmpersandAmpersandToken) {
-            const match = checkMatch(matchTo, expression.right, checker);
+            const match = getMatchingSubexpression(matchTo, expression.right, checker);
             if (!match) {
                 break;
             }
             occurrences.push(match);
             matchTo = match;
             expression = expression.left;
         }
-        const finalMatch = checkMatch(matchTo, expression, checker);
+        const finalMatch = getMatchingSubexpression(matchTo, expression, checker);
         if (finalMatch) {
             occurrences.push(finalMatch);
         }
         return occurrences.length > 0 ? occurrences: undefined;
     }
 
     /**
-     * Checks that expression is a subchain of matchTo.
+     * Checks that expression is a syntactic subexpression of matchTo.
      */
-    function checkMatch(matchTo: Expression, expression: Expression, checker: TypeChecker): PropertyAccessExpression | Identifier | undefined {
+    function getMatchingSubexpression(matchTo: Expression, expression: Expression, checker: TypeChecker): PropertyAccessExpression | Identifier | undefined {
         if (isCallExpression(matchTo)) {
-            // Recurse through the call expressions to match a.b to a.b()().
-            return !isCallExpression(expression) ? checkMatch(matchTo.expression, expression, checker) : undefined;
+            return !isCallExpression(expression) ? getMatchingSubexpression(matchTo.expression, expression, checker) : undefined;
         }
         else if ((isPropertyAccessExpression(expression) || isIdentifier(expression)) && (isPropertyAccessExpression(matchTo) || isIdentifier(matchTo))) {
-            return checker.isOrContainsMatchingReference(matchTo, expression) ? expression : undefined;
+            return containsSyntacticSubchain(matchTo, expression, checker) ? expression : undefined;
         }
         return undefined;
     }
 
+    /**
+     * Returns true if target is a syntactic subchain of source.
+     */
+    function containsSyntacticSubchain(source: Node, target: Node, checker: TypeChecker): boolean {
+        while (isPropertyAccessExpression(source) && !isSyntacticMatch(source, target)) {
+            source = source.expression;
+        }
+        while (isPropertyAccessExpression(source) && isPropertyAccessExpression(target)) {
+            if (!isSyntacticMatch(source, target)) return false;
+            source = source.expression;
+            target = target.expression;
+        }
+        const fullMatch = isSyntacticMatch(source, target);
+        if (fullMatch && !isInJSFile(source)) {
+            Debug.assert(checker.getSymbolAtLocation(source) === checker.getSymbolAtLocation(target));
+        }
+        return fullMatch;

```

Comment:
I think there’s possibly a less efficient but more easily understandable way to write this function by going depth-first, grabbing the left-most name of both the chain and the subchain right away, and walking up their parents. But I couldn’t write an elegant example in a few minutes, so maybe it wouldn’t be as good as I think it would.

Reply:
I guess the easiest to understand way to break it down, in my mind, would be to have a function that returns the array `["a", "b", "c", "d"]` for the PropertyAccessExpression `a.b.c.d` (it could return strings or Identifiers, either way), then

```ts
function chainStartsWith(chain: EntityNameExpression, subchain: EntityNameExpression) {
  const subchainIdentifiers = getIdentifiersInEntityName(subchain);
  const chainIdentifiers = getIdentifiersInEntityName(chain);
  return every(subchain, (identifier, index) => {
    return identifier.getText() === chainIdentifiers[index]?.getText();
  });
}
```

This would be a bit more eager than what you have, and would allocate some arrays you don’t strictly need, but is easier to follow to me.

Reply:
In this example I also presuppose the types `EntityNameExpression` to guarantee that you have a string of dot-separated identifiers, because it makes this function easier to reason about and there are already `isEntityNameExpression` functions you can use, but if you wanted to nitpick on efficiency, you could check on the fly in `getIdentifiersInEntityName`, returning `undefined` if you discover that you don’t actually have an EntityNameExpression.

Reply:
Would this make other cases we care about easier to implement (maybe optional element access) in addition to making it easier to read? It does feel quite awkward to go through the entirety of both access chains to build the arrays if we would terminate on the first check we would have made otherwise. On the other hand, really long access chains in binary expressions are probably not common anyway?

Reply:
I think expanding to optional element access should be pretty easy either way.